[1.18.x] Hitchhiker's Guide to All Things Recipes (#420)
MinecraftForge/Documentation: docs/resources/server/recipes/index.md

Recipes
=======

Recipes are a way to transform some number of objects into other objects within a Minecraft world. Although the vanilla system deals purely with item transformations, the system as a whole can be expanded to use any object the programmer creates.

Data-Driven Recipes
-------------------

Most recipe implementations within vanilla are data driven via JSON. This means that a mod is not necessary to create a new recipe, only a [Data pack][datapack]. A full list on how to create and put these recipes within the mod's `resources` folder can be found on the [Minecraft Wiki][wiki].

> A recipe can be obtained within the Recipe Book as a reward for completing an [advancement][advancement]. Recipe advancements always parent `minecraft:recipes/root` so as not to appear on the Advancement screen and that they are treated as standalone. In addition, the criteria to unlock a recipe always checks if the user has unlocked the recipe from using it once:

Reviewer: The wording here is a bit awkward to me. Also, what does it mean for an advancement to be "treated as standalone"? (I omitted that part from my suggestion -- feel free to re-add it back with a sufficient explanation.)
```suggestion
A recipe can be obtained within the Recipe Book as a reward for completing an [advancement][advancement]. Recipe advancements always have `minecraft:recipes/root` as their parent, to not to appear on the advancement screen. The default criteria to gain the recipe advancement is a check if the user has unlocked the recipe from using it once or receiving it through a command like `/recipe`:
```

```json5
// Within some recipe advancement json
"has_the_recipe": { // Criteria label
  // Succeeds if examplemod:example_recipe is used
  "trigger": "minecraft:recipe_unlocked",
  "conditions": {
    "recipe": "examplemod:example_recipe"
  }
}
//...
"requirements": [
  [
    "has_the_recipe"
    // ... Other criteria labels to be ORed against to unlock recipe
  ]
]
```

Data-driven recipes and their unlocking advancement can be [generated][datagen] via `RecipeProvider`.

Recipe Manager
--------------

Recipes are loaded and stored via the `RecipeManager`. Any operations relating to getting available recipe(s) are handled by this manager. There are two important methods to know of:

Method|Description
:---: | :---
getRecipeFor | Gets the first recipe that matches the current input.

> getRecipesFor | Gets all recipes that match the current input.

Reviewer: Misalignment on the column separators. Also, we may want to have the method names in monospace.
```suggestion
 Method         | Description
 :---:          | :---
`getRecipeFor`  | Gets the first recipe that matches the current input.
`getRecipesFor` | Gets all recipes that match the current input.
```

Each method takes in a `RecipeType`, which denotes what method is being applied to use the recipe (crafting, smelting, etc.), a `Container` which holds the configuration of the inputs, and the current level which is passed to `Recipe#matches` along with the container.

!!! important

>     When using an [item handler capability][cap] as the configuration inputs, it must be wrapped via `RecipeWrapper` and passed in as `IItemHandler` and its subtypes are not instances of `Container`:

Reviewer: I think this can be worded more helpfully by reducing the imperative command of "must be wrapped" into mentioning the use of `RecipeWrapper` for wrapping `IItemHandler`s:
```suggestion
    Forge provides the `RecipeWrapper` utility class which extends `Container` for wrapping around `IItemHandler`s and passing them to methods which requires a `Container` parameter.
```

    ```java
    // Within some method with IItemHandlerModifiable handler
    recipeManger.getRecipeFor(RecipeType.CRAFTING, new RecipeWrapper(handler), level);
    ```

Additional Features
-------------------

Forge provides some additional behavior to the recipe schema and its implementations for greater control of the system.

### Recipe ItemStack Result

> Except for `minecraft:stonecutting` recipes, all recipe serializers expand the `result` tag to take in a full `ItemStack` as a `JsonObject` instead of just the item name and amount in some cases.

Reviewer: All recipe serializers, or all _vanilla_ recipe serializers?

```json5
// In some recipe JSON
"result": {
  // The name of the registry item to give as a result
  "item": "examplemod:example_item",
  // The number of items to return
  "count": 4,
  // The tag data of the stack, can also be a string
  "nbt": {
      // Add tag data here

>   }

Reviewer: We should mention somehow that the `nbt` tag can either be a JSON object, or a string containing SNBT (stringified NBT) for NBT which cannot be represented as a JSON object (such as those involving `IntArrayTag`s).
